model/build/build.go

```diff
@@ -59,8 +59,36 @@ type Build struct {
 
 // Returns whether or not the build has finished, based on its status.
 func (b *Build) IsFinished() bool {
-	return b.Status == evergreen.BuildFailed ||
-		b.Status == evergreen.BuildSucceeded
+	return b.Status == evergreen.BuildSucceeded ||
+		(b.Status == evergreen.BuildFailed && b.allCachedTasksFinished())
+}
+
+func taskCacheTaskIsUnscheduled(t *TaskCache) bool {
+	if !t.Activated && t.Status == evergreen.TaskUndispatched {
+		return true
+	}
+
+	return false
+}
+
+// allCachedTasksFinished returns true when either:
+//  1. if there is a compile task, the compile task's status is one the ones
+//     listed in IsFinishedTaskStatus
+//  2. or the task's status is one of the finished tasks listed in IsFinishedTaskStatus
+func (b *Build) allCachedTasksFinished() bool {
+	allFinished := true
+	for i := range b.Tasks {
+		if !taskCacheTaskIsUnscheduled(&b.Tasks[i]) && !evergreen.IsFinishedTaskStatus(b.Tasks[i].Status) {
```

Comment:
What state are you capturing here, with (is not unscheduled) and (is not finished)?